### PR TITLE
replace builtin DS backend with builtin_local and builtin_raft

### PR DESCRIPTION
**Pull request by ieQu1**
Fixes EMQX-12491

Release version: v/e5.8

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [x] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_meta.erl [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+-spec set_current_timestamp(emqx_ds_storage_layer:shard_id(), emqx_ds:time()) -> ok.
+set_current_timestamp(ShardId, Time) ->
+    mria:dirty_write(?TS_TAB, #?TS_TAB{id = ShardId, latest = Time}).
+
+%% Return a timestamp that is monotonic between the node restarts.
+%%
+%% This property is maintained using the following mechanism:
+%% - `local_meta' gen_server maintains two values as `atomics':
+%%    - `offset':
+%%
```
Comment (ieQu1) [inline apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_meta.erl:141]: Remove me, it was part of an idea that was ultimately rejected.

File: apps/emqx/src/emqx_ds_schema.erl [hunk 2]
```diff
@@ -32,42 +32,51 @@
 %% Type declarations
 %%================================================================================
 
+-ifndef(EMQX_RELEASE_EDITION).
+-define(EMQX_RELEASE_EDITION, ce).
```
Comment (thalesmg) [inline apps/emqx/src/emqx_ds_schema.erl:36]: Just curious: why was this needed?  It should have been set by rebar/mix during compilation.  :thinking:

File: apps/emqx/test/emqx_persistent_messages_SUITE.erl [hunk 3]
```diff
@@ -32,9 +32,23 @@
 all() ->
     emqx_common_test_helpers:all(?MODULE).
 
+%% Needed for standalone mode:
```
Comment (thalesmg) [inline apps/emqx/test/emqx_persistent_messages_SUITE.erl:35]: What is "standalone mode"?

File: apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_app.erl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+%% behavior callbacks:
+-export([start/2]).
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+start(_StartType, _StartArgs) ->
+    emqx_ds:register_backend(builtin_local, emqx_ds_builtin_local),
+    %% TODO: fixme
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_app.erl:30]: Yes, well spotted. This `self()` trick is dubious.

File: apps/emqx/src/emqx_ds_schema.erl (hunk 2 above)
Comment (ieQu1) [reply apps/emqx/src/emqx_ds_schema.erl:36]: Once again, it's because of the standalone mode.

File: apps/emqx_ds_builtin_raft/src/emqx_ds_builtin_raft_app.erl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_ds_builtin_raft_app).
+
+-export([start/2]).
+
+start(_Type, _Args) ->
+    emqx_ds:register_backend(builtin_raft, emqx_ds_replication_layer),
```
Comment (thalesmg) [inline apps/emqx_ds_builtin_raft/src/emqx_ds_builtin_raft_app.erl:10]: Should we deregister on stop?

File: apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_app.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local_app.erl]: Fixed. Now backend supervisors live in their own apps.

File: changes/ce/breaking-13248.en.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+`builtin` durable storage backend has been replaced with the following two backends:
+
+- `builtin_local`: A durable storage backend that doesn't support replication.
+   It can't be used in a multi-node cluster.
+   This backend is available in both open source and enterprise editions.
+- `builtin_raft`: A durable storage backend that uses Raft algorithm for replication.
+   This backend is available enterprise edition.
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply changes/ce/breaking-13248.en.md]: Thanks, fixed.

File: apps/emqx/src/emqx_ds_schema.erl (hunk 2 above)
Comment (thalesmg) [reply apps/emqx/src/emqx_ds_schema.erl:36]: Aha, when testing just the `emqx` app, I see.

File: apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    ).
+
+-spec list_generations_with_lifetimes(emqx_ds:db()) ->
+    #{emqx_ds:generation_rank() => emqx_ds:generation_info()}.
+list_generations_with_lifetimes(DB) ->
+    lists:foldl(
+        fun(Shard, Acc) ->
+            maps:fold(
+                fun(GenId, Data, Acc1) ->
+                    Acc1#{{Shard, GenId} => Data}
```
Comment (keynslug) [inline apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl:159]: Since `emqx_persistent_message_ds_gc_worker` currently expects lifetime-related timestamps are in milliseconds, this will break those expectations AFAICT. Probably either timestamps need to be converted back, or callers taught to deal with finer timestamps?

File: apps/emqx_durable_storage/src/emqx_ds_buffer.erl [hunk 8]
```diff
@@ -115,18 +133,22 @@ store_batch(DB, Messages, Opts) ->
...
+    {ok, CallbackS} = CBM:init_buffer(DB, Shard, CBMOptions),
     S = #s{
+        callback_module = CBM,
+        callback_state = CallbackS,
         db = DB,
         shard = Shard,
         metrics_id = MetricsId,
         queue = queue:new()
     },
+    persistent_term:put(?cbm(DB), {CBM, CBMOptions}),
```
Comment (keynslug) [inline apps/emqx_durable_storage/src/emqx_ds_buffer.erl:151]: Nit. TBH having shard-level processes write to DB-level persistent term feels a bit awkward. I wonder why this module can't pull this CBM info from `emqx_ds` or `emqx_ds_sup`.

**Review by keynslug — COMMENTED** [review]
Great work!

File: apps/emqx_ds_backends/src/emqx_ds_backends.app.src.script [hunk 9]
```diff
@@ -0,0 +1,26 @@
+%% -*- mode: erlang -*-
+Profile = case os:getenv("PROFILE") of
+            "emqx-enterprise" ++ _ ->
+              ee;
+            false ->
+              io:format(user, "WARN: environment variable PROFILE is not set, using 'emqx-enterprise'~n", []),
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_ds_backends/src/emqx_ds_backends.app.src.script:6]: Yes. But I faced some non-trivial problems with inconsistent handling of the default profile. 
Printing this warning every time we assume the default profile will help to investigate the inconsistencies in the future.

File: apps/emqx/src/emqx_ds_schema.erl [hunk 10]
```diff
@@ -252,10 +247,55 @@ fields(layout_builtin_reference) ->
...
+        {n_shards,
+            sc(
+                pos_integer(),
+                #{
+                    default => 12,
+                    importance => ?IMPORTANCE_MEDIUM,
+                    desc => ?DESC(builtin_n_shards)
+                }
+            )},
+        {local_write_buffer,
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx/src/emqx_ds_schema.erl:271]: True. But I didn't want to deviate config too much.

File: apps/emqx_durable_storage/src/emqx_ds_buffer.erl (hunk 8 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_ds_buffer.erl:151]: `emqx_ds` ideally should not be aware of any specifics of the backend, including whether or not it uses the buffer.

File: apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl (hunk 7 above)
Comment (ieQu1) [reply apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl:164]: Good catch, fixed.
P.S. That was one of the reasons I was against unit conversions in DS :slightly_smiling_face:

File: apps/emqx_durable_storage/src/emqx_ds_buffer.erl [hunk 11]
```diff
@@ -246,39 +271,41 @@ do_flush(
...
+            emqx_ds_builtin_metrics:inc_buffer_bytes(Metrics, S#s.n_bytes),
             ?tp(
-                emqx_ds_replication_layer_egress_flush,
+                emqx_ds_buffer_flush,
                 #{db => DB, shard => Shard, batch => Messages}
             ),
             lists:foreach(fun(From) -> gen_server:reply(From, ok) end, Replies),
             erlang:garbage_collect(),
             S#s{
+                callback_state = CallbackS,
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_ds_buffer.erl]: Indeed, fixed.

**Review by thalesmg — APPROVED** [review]
I haven't  finished reviewing, but approving so it unblocks further work and can be fixed in follow ups.

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3402)
